internal/linker/linker.go

```diff
@@ -0,0 +1,263 @@
+			return nil, err
+		}
+
+		if err := applyPatch(workingDirectory, patch); err != nil {
+			return nil, fmt.Errorf("apply patch for %s failed: %v", fileName, err)
+		}
+	}
+	return mod, nil
+}
+
+func cachePath(goExe string) (string, error) {
+	var cacheDir string
+	if val, ok := os.LookupEnv(garbleCacheDir); ok {
+		cacheDir = val
+	} else {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			panic(fmt.Errorf("cannot retreive user cache directory: %v", err))
+		}
+		cacheDir = userCacheDir
+	}
+
+	cacheDir = filepath.Join(cacheDir, cacheDirName)
+	if err := os.MkdirAll(cacheDir, os.ModePerm); err != nil {
+		return "", err
+	}
+
+	return filepath.Join(cacheDir, "link"+goExe), nil
+}
+
+func getCurrentVersion(goVersion string) string {
+	return linkerPatchesVer + " " + goVersion
+}
+
+func checkVersion(linkerPath, goVersion string) (bool, error) {
+	versionPath := linkerPath + versionExt
+	version, err := os.ReadFile(versionPath)
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+
+	return string(version) == getCurrentVersion(goVersion), nil
+}
+
+func writeVersion(linkerPath, goVersion string) error {
+	versionPath := linkerPath + versionExt
+	return os.WriteFile(versionPath, []byte(getCurrentVersion(goVersion)), os.ModePerm)
+}
+
+type overlayFile struct {
+	Replace map[string]string
+}
+
+func compileLinker(workingDirectory string, overlay map[string]string, outputLinkPath string) error {
+	file, err := json.Marshal(&overlayFile{Replace: overlay})
+	if err != nil {
+		return err
+	}
+	overlayPath := filepath.Join(workingDirectory, "overlay.json")
+	if err := os.WriteFile(overlayPath, file, os.ModePerm); err != nil {
+		return err
+	}
+
+	cmd := exec.Command("go", "build", "-overlay", overlayPath, "-o", outputLinkPath, "cmd/link")
+	// Explicitly setting GOOS and GOARCH variables prevents conflicts during cross-build
+	cmd.Env = append(os.Environ(), "GOOS="+runtime.GOOS, "GOARCH="+runtime.GOARCH)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("compiler compile error: %v\n\n%s", err, string(out))
+	}
+
+	return nil
+}
+
+func GetModifiedLinker(goRoot, goVersion, goExe, tempDirectory string) (string, error) {
+	outputLinkPath, err := cachePath(goExe)
+	if err != nil {
+		return "", err
+	}
+
+	// Used for double check. Quick check before filelock and safe check after filelock.
+	checkLinker := func() (bool, error) {
+		isCorrectVer, err := checkVersion(outputLinkPath, goVersion)
+		if err != nil {
+			return false, err
+		}
+		return isCorrectVer && existsFile(outputLinkPath), nil
+	}
+
+	isCorrectVer, err := checkLinker()
+	if isCorrectVer || err != nil {
+		return outputLinkPath, err
+	}
+
+	fileLock := flock.New(outputLinkPath + ".lock")
+	if err := fileLock.Lock(); err != nil {
+		return "", err
+	}
+	defer fileLock.Unlock()
+
+	isCorrectVer, err = checkLinker()
+	if isCorrectVer || err != nil {
+		return outputLinkPath, err
+	}
```

Comment:
I'm afraid this is buggy. Since you only ever keep one patched linker binary in the cache, you can have this sequence of events:

1) Start with an empty cache; no previously patched linker
2) Start `garble build` with garble version X with Go version Y1; let's call this build process P1
3) P1 stores the cached linker with its version, and finishes the build
4) Start (another) `garble build` with garble version X with Go version Y1; let's call this build process P1 (again)
5) Start `garble build` with garble version X with Go version Y2; let's call this build process P2
4) P1 notices that the cached linker is present with its version; it wants to use the cached path.
5) Concurrently (P1 and P2 are concurrent), P2 sees the cached linker is present, but the Go version is a mismatch.
6) P2 grabs the lock to rebuild the linker and replaces P1's cached linker with P2.
7) If P1 continues the build now by executing the cached linker, it is running P2's linker, and not P1's.

The simplest way to solve this would be to grab the file lock before the first `checkLinker` call, and release the lock _after_ the entire linker execution is complete. Otherwise, the linker file might be replaced before it's actually run.

On Linux, you could probably hold the lock file for less time by releasing once https://pkg.go.dev/os/exec#Cmd.Start has finished - by that point, I assume that the binary has been loaded and the process started, so we don't need to wait for the linking to finish to be able to delete or replace the file. However, that may cause problems on Windows, where a file being executed cannot be safely deleted or modified, if I recall correctly.

At a later time, I think we can improve all of this by using GOCACHE's own implementation for our cache, where the "version" would be the cache key. That would allow us to store multiple linker binaries at once, and it would deal with cleaning up old files without causing "deleted while being used" problems on Windows. But for now, I think holding the lock for as long as the file is being used is OK.